packages/expo-doctor/src/utils/explainDependencies.ts

```diff
@@ -0,0 +1,132 @@
+import spawnAsync, { SpawnResult } from '@expo/spawn-async';
+import chalk from 'chalk';
+import semver from 'semver';
+
+import { RootNodePackage, VersionSpec } from './explainDependencies.types';
+
+type TargetPackage = { name: string; version?: VersionSpec };
+
+function isSpawnResult(result: any): result is SpawnResult {
+  return 'stderr' in result && 'stdout' in result && 'status' in result;
+}
+
+/** Spawn `npm explain [name] --json` and return the parsed JSON. Returns `null` if the requested package is not installed. */
+async function explainAsync(
+  packageName: string,
+  projectRoot: string,
+  parameters: string[] = []
+): Promise<RootNodePackage[] | null> {
+  const args = ['explain', packageName, ...parameters, '--json'];
+
+  try {
+    const { stdout } = await spawnAsync('npm', args, {
+      stdio: 'pipe',
+      cwd: projectRoot,
+    });
+
+    return JSON.parse(stdout);
+  } catch (error: any) {
+    if (isSpawnResult(error)) {
+      if (error.stderr.match(/npm ERR! No dependencies found matching/)) {
+        return null;
+      } else if (error.stdout.match(/Usage: npm <command>/)) {
+        throw new Error(
+          `Dependency tree validation for ${chalk.underline(
+            packageName
+          )} failed. This validation is only available on Node 16+ / npm 8.`
+        );
+      }
+    }
+    throw new Error(`Failed to find dependency tree for ${packageName}: ` + error.message);
+  }
+}
+
+function organizeExplanations(
+  pkg: TargetPackage,
+  {
+    explanations,
+    isValid,
+  }: {
+    explanations: RootNodePackage[];
+    isValid: (pkg: TargetPackage) => boolean;
+  }
+) {
+  const valid: RootNodePackage[] = [];
+  const invalid: RootNodePackage[] = [];
+
+  for (const explanation of explanations) {
+    const { name } = explanation;
+    if (name === pkg.name) {
+      if (isValid(explanation)) {
+        valid.push(explanation);
+      } else {
+        invalid.push(explanation);
+      }
+    }
+  }
+  return { valid, invalid };
+}
+
+/**
+ * @param pkg
+ * @param explanations
+ * @returns true if all versions of the package satisfy the constraints
+ */
+async function printExplanationsAsync(
```

Comment:
I do not expect a method called `print...` to also return something, in similar situations in the CLIs we usually split this up so there's no mixed concerns.

Reply:
This bubbles up as well. The function that calls this (warnAbout....) also combines printing output (by calling into this method) and bubbling up a return value. This to me was one of the biggest issues with the old implementation. I would seriously consider refactoring this to achieve a clean separation between logic/flow and output.